compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6970,6 +6970,20 @@ private BType checkFieldAccessExpr(BLangFieldBasedAccess fieldAccessExpr, BType
             actualType = checkObjectFieldAccessExpr(fieldAccessExpr, varRefType, fieldName);
             fieldAccessExpr.originalType = actualType;
         } else if (types.isSubTypeOfBaseType(varRefType, TypeTags.RECORD)) {
+
+            // Check if the field accessed is present in the record fields set.
+            if (varRefType.tag == TypeTags.RECORD && !checkRecordFieldExistence((BRecordType) varRefType, fieldName)) {
+                dlog.error(fieldAccessExpr.pos, DiagnosticErrorCode.UNDEFINED_FIELD_IN_RECORD, fieldName, varRefType);
+                return actualType;
+            }
+
+            // Check if the field accessed is present in the fields of a record which is a member of a union type.
+            if (varRefType.tag == TypeTags.UNION &&
+                    !checkFieldExistenceInUnionRecordMembers(((BUnionType) varRefType).getMemberTypes(), fieldName)) {
+                dlog.error(fieldAccessExpr.pos, DiagnosticErrorCode.UNDEFINED_FIELD_IN_RECORD, fieldName, varRefType);
+                return actualType;
+            }
```

Comment:
Is this the best way to handle this? These 2 fixes placed here seem rather ad-hoc. The logic related to this is in the `checkRecordFieldAccessExpr()` method. Shouldn't this logic be incorporated somewhere in that method?

Reply:
I also think we can improve how we handle this. But `checkRecordFieldAccessExpr` doesn't log errors. I think we can let this go on the original path and handle the errors from L6994 onward.

If we do it this way we'll be doing some checks unnecessarily. For example where the field is valid, both the introduced checks and  `checkRecordFieldAccessExpr` will check for existence.

Reply:
Basically if `!fieldAccessExpr.isLValue` iin L6994, do these checks and use the relevant message.

Reply:
Updated the approach.